File: example/flux_example_loan/flux_example_loan/image_pub.py
#!/usr/bin/env python3
"""Publish an Image into a loaned slot, with no copy.

ros2 run flux_example_loan image_pub.py
"""

import flux.ros
import numpy as np
import rclpy
from rclpy.executors import ExternalShutdownException
from rclpy.node import Node
from sensor_msgs_flux.image import Image


class ImageLoanPublisher(Node):

    TOPIC = "image"
    FRAME_ID = "camera"
    ENCODING = "bgr8"
    WIDTH = 1920
    HEIGHT = 1200
    STEP = WIDTH * 3
    DATA_BYTES = STEP * HEIGHT
    SLOT_SIZE = DATA_BYTES + 1024
    SLOT_COUNT = 8
    RATE_HZ = 30.0

    def __init__(self):
        super().__init__("flux_loan_image_pub")
        self.ramp = np.arange(self.DATA_BYTES, dtype=np.uint8)
        self.phase = 0
        self.publisher = flux.ros.Publisher(
            self,
            self.TOPIC,
            fingerprint=Image.FINGERPRINT__,
            slot_size=self.SLOT_SIZE,
            slot_count=self.SLOT_COUNT,
        )
        self.timer = self.create_timer(1.0 / self.RATE_HZ, self.tick)

    def tick(self):
        b = Image.build__(self.publisher)
        if b is None:
            return
        stamp = self.get_clock().now().nanoseconds
        b.set__header__stamp(stamp // 1000000000, stamp % 1000000000)
        b.header__frame_id = self.FRAME_ID
        b.height = self.HEIGHT
        b.width = self.WIDTH
        b.encoding = self.ENCODING
        b.is_bigendian = 0
        b.step = self.STEP
        np.add(self.ramp, np.uint8(self.phase), out=b.alloc__data(self.DATA_BYTES))
        self.report(b.commit__())
        self.phase = (self.phase + 1) % 256

    def report(self, published):
        """Backpressure is a dropped frame on a best-effort transport, so it is not an error.
        Anything else does not clear on its own."""
        if flux.faulted(published):
            self.get_logger().error(f"publish refused: {published}")


def main(args=None):
    rclpy.init(args=args)
    node = ImageLoanPublisher()
    try:
        rclpy.spin(node)
    except (KeyboardInterrupt, ExternalShutdownException):
        pass
    finally:
        node.destroy_node()
        rclpy.try_shutdown()


if __name__ == "__main__":
    main()

File: example/flux_example_loan/src/image_pub.cpp
#include "flux/ros/publisher.hpp"

#include <rclcpp/rclcpp.hpp>

#include "sensor_msgs/flux/image.hpp"

#include <chrono>
#include <cstddef>
#include <cstdint>
#include <memory>

class ImageLoanPublisher : public rclcpp::Node
{
public:
  ImageLoanPublisher()
  : Node("flux_loan_image_pub"), pub_(*this, kTopic, Image::kFingerprint, kSlotSize, kSlotCount)
  {
    timer_ = create_wall_timer(
      std::chrono::duration_cast<std::chrono::nanoseconds>(
        std::chrono::duration<double>(1.0 / kRateHz)),
      [this]() { tick(); });
  }

private:
  using Image = sensor_msgs::flux_msg::Image;

  static constexpr const char * kTopic = "image";
  static constexpr const char * kFrameId = "camera";
  static constexpr const char * kEncoding = "bgr8";
  static constexpr std::uint32_t kWidth = 1920;
  static constexpr std::uint32_t kHeight = 1200;
  static constexpr std::uint32_t kStep = kWidth * 3;
  static constexpr double kRateHz = 30.0;
  static constexpr std::uint32_t kSlotCount = 8;
  static constexpr std::size_t kDataBytes = static_cast<std::size_t>(kStep) * kHeight;
  static constexpr std::uint32_t kSlotSize = static_cast<std::uint32_t>(kDataBytes + 1024);

  void tick()
  {
    Image::Builder b = Image::build__(pub_);
    if (!b) {
      return;  // every slot borrowed; delivery is best-effort
    }

    const rclcpp::Time t = now();
    b.set__header__stamp(static_cast<std::int32_t>(t.seconds()), t.nanoseconds() % 1000000000u);
    b.set__header__frame_id(kFrameId);
    b.set__height(kHeight);
    b.set__width(kWidth);
    b.set__encoding(kEncoding);
    b.set__is_bigendian(0);
    b.set__step(kStep);

    auto data = b.alloc__data(kDataBytes);
    for (std::size_t i = 0; i < data.size(); ++i) {
      data[i] = static_cast<std::uint8_t>(i + phase_);
    }
    // Backpressure is a dropped frame on a best-effort transport, so it is not an error.
    // Anything else does not clear on its own.
    if (const flux::Published p = b.commit__(); flux::faulted(p)) {
      RCLCPP_ERROR_ONCE(get_logger(), "commit refused: %s", flux::to_string(p));
    }
    ++phase_;
  }

  flux::ros::Publisher pub_;
  rclcpp::TimerBase::SharedPtr timer_;
  std::uint8_t phase_ = 0;
};

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<ImageLoanPublisher>());
  rclcpp::shutdown();
  return 0;
}
